jax/_src/api.py

```diff
@@ -1045,8 +1030,8 @@ def jacfun(*args, **kwargs):
     jac = vmap(pullback)(_std_basis(y))
     jac = jac[0] if isinstance(argnums, int) else jac
     example_args = dyn_args[0] if isinstance(argnums, int) else dyn_args
-    jac = tree_map(partial(_unravel_array_into_pytree, y, 0), jac)
-    return tree_transpose(tree_structure(example_args), tree_structure(y), jac)
+    jac_tree = tree_multimap(partial(_jacrev_unravel, y), example_args, jac)
```

Comment:
Oh nice, I didn't know that!  Should I switch to using `tree_map` exclusively in my code?

Reply:
You certainly could. We haven't deprecated `tree_multimap`, so either is fine, but `tree_map` is shorter!

Reply:
> To what extent does the original issue remain unsolved if the caller were to tree-map a cast-to-complex function before calling `jacfwd`/`jacrev`? 

That's already what JAX does.  The `_std_basis` function chooses as its type the `tree_reduce` using `result_type` of the input tree.  So if there are any complex components, it chooses a basis of complex type.

In my code, this is a problem [here, for example](https://github.com/NeilGirdhar/efax/blob/master/efax/_src/distributions/complex_normal.py#L28).  I want to be able to calculate the _Jacobian of the gradient_.  The function expects `self.precision` to be real-valued, and must produce a real-valued output.  If we cast the inputs to complex, then I need to replace all uses of `self.precision` with `self.precision.real` to discard the complex components.  If I don't do that, I'll get an exception in JAX because the output will be complex, which the gradient doesn't support.

I think it's surprising that a function that's called with real-valued `self.precision` should be called with complex values by JAX.  Even if the XLA compiler optimizes it away, I think it's computationally wasteful to pass in synthetic complex components that are ultimately discarded.

And this isn't just a problem with complex types.  Imagine if you had a function with one input component that's `np.float64` and one that's `np.float32`.  Wouldn't it be nice if Jacobian should respected that and passed in the same types for its calculation?  Currently, the narrower type is being silently upcast.

Reply:
Thanks for clarifying. If I understand correctly: we're arguably doing the surprising thing already, and either way, casting to complex can confuse a function that was expecting some strictly real-valued arguments.

@mattjj, do you recall whether we had a reason to unify the tangent types when forming the standard basis for `jacfwd`/`jacrev`?

By the way, should the tests in this PR cover the case where "the narrower type is silently widened" in a way that requires the function implementation to change, as in having to write `x.real`?

Reply:
> By the way, should the tests in this PR cover the case where "the narrower type is silently widened" in a way that requires the function implementation to change, as in having to write x.real?

Good point. I will add that test.

Reply:
Done adding the test.  Please let me know if you need me to do anything else.